### PR TITLE
Add API to annotate disjoint backward and handle in AC

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #167251
* #167006
* #166610
* __->__ #166536

This adds zero-bubble / DualPipeV support for (S)AC

Before:
- AC will always retrigger recompute upon every distinct backward.

After:
- Any checkpointed regions encountered by backward under the same instance of this context manager will only trigger recompute at most once, even if there are multiple calls to backward.
- Backward calls under the same instance of this context manager must execute over non-overlapping regions of the backward graph even if retain_graph=True.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

### CODE REVIEW

File: torch/autograd/graph.py [hunk 1]
```diff
@@ -254,6 +255,39 @@ def increment_version(tensor: Union[torch.Tensor, Iterable[torch.Tensor]]) -> No
     torch._C._increment_version(tensor)
 
 
+class GraphExecutionGroup:
```
Comment (soulitzer) [inline torch/autograd/graph.py]: Name bikesheddable, could be something more related to disjointness

File: torch/utils/checkpoint.py [hunk 2]
```diff
@@ -1587,6 +1592,40 @@ def recompute_fn(*inputs):
 
     return
 
+
+class GraphExecGroup:
+    """Any checkpointed regions encountered by backward under the same instance
+    of this context manager will only trigger recompute at most once, even if
```
Comment (albanD) [inline torch/utils/checkpoint.py]:
```suggestion
    of this context manager will trigger recompute at most once, even if
```

File: torch/utils/checkpoint.py [hunk 3]
```diff
@@ -1587,6 +1592,40 @@ def recompute_fn(*inputs):
...
+    there are multiple calls to backward.
+
+    Backward calls under the same instance of this context manager must execute
+    over non-overlapping regions of the backward graph even if retain_graph=True.
```
Comment (albanD) [inline torch/utils/checkpoint.py:1616]: Why? If they don't overlap, then they would never trigger the recompute more than once, hence according to the paragraph above, this context manager is a noop? :D

File: torch/utils/checkpoint.py [hunk 4]
```diff
@@ -1587,6 +1592,40 @@ def recompute_fn(*inputs):
 
     return
 
+
+class GraphExecGroup:
```
Comment (albanD) [inline torch/utils/checkpoint.py:1610]: nit: for the name, feel free to ignore, but why are we talking of "graph" here?
Also we should mention that this is only relevant for non-reentrant checkpoint

File: test/test_autograd.py [hunk 5]
```diff
@@ -7362,6 +7362,61 @@ def test_checkpoint_sequential_warns_if_use_reentrant_not_passed_explcitly(self)
...
+
+        # Test the not actually disjoint case (using retain_graph=True since
+        # otherwise autograd itself will catch this)
+        def fn(x):
+            return x.sin().cos()
+
+        x = torch.randn(3, 3, requires_grad=True)
+        out = checkpoint(fn, x, use_reentrant=False)
+        group = torch.utils.checkpoint.GraphExecGroup()
+        with group:
```
Comment (albanD) [inline test/test_autograd.py]: Note that usually in python you would inline the context manager creation there.

File: torch/utils/checkpoint.py [hunk 6]
```diff
@@ -1587,6 +1592,40 @@ def recompute_fn(*inputs):
...
+
+    Backward calls under the same instance of this context manager must execute
+    over non-overlapping regions of the backward graph even if retain_graph=True.
+    """
+    _current_group: Optional[int] = None
+    _counter: int = 0
+
+    def __init__(self) -> None:
+        self._id: int = GraphExecGroup._counter
+        GraphExecGroup._counter += 1
```
Comment (albanD) [inline torch/utils/checkpoint.py]: is there a thread safe version of this?

File: torch/utils/checkpoint.py [hunk 7]
```diff
@@ -1587,6 +1592,40 @@ def recompute_fn(*inputs):
...
+    of this context manager will only trigger recompute at most once, even if
+    there are multiple calls to backward.
+
+    Backward calls under the same instance of this context manager must execute
+    over non-overlapping regions of the backward graph even if retain_graph=True.
+    """
+    _current_group: Optional[int] = None
+    _counter: int = 0
+
+    def __init__(self) -> None:
```
Comment (albanD) [inline torch/utils/checkpoint.py]: Since we never expect this to be set globally, you most likely can make this a regular context manager with something like https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager that will make your life a lot easier

File: torch/utils/checkpoint.py (hunk 3 above)
Comment (soulitzer) [reply torch/utils/checkpoint.py:1616]: I think there's two concepts of overlap here:
- two backward passes resulting in unpacking the same saved activation
- two backward passes unpacking two different saved activation, but those two saved activation were saved while under the same checkpoint region.

In this description I mean that we don't allow the first kind of overlap! 
This context manager is needed due to the second kind of overlap, which will cause recompute can be triggered twice.

File: test/test_autograd.py (hunk 5 above)
Comment (soulitzer) [reply test/test_autograd.py]: Yup, although I think I also wanted to emphasize that the incrementing is done in the construction, and I'm able to use it in this manner (but I forgot to do so in this test!)
```
group = ...
with group:
    ...
with group:
    ...
```

File: torch/utils/checkpoint.py (hunk 4 above)
Comment (soulitzer) [reply torch/utils/checkpoint.py:1610]: Not super opinionated on this, but graph exec group seemed analogous to graph task id, and it's a group of (disjoint) backward graph execution.

Good catch, let me mention non-reentrant in the description.

File: torch/utils/checkpoint.py (hunk 7 above)
Comment (soulitzer) [reply torch/utils/checkpoint.py]: IIUC if we use that, each instance of that one can only be entered/exited only one time?
I might prefer to have the ability to create the object representing the group once, and then wrap any .grads individually. This would be useful when I'm implementing pipelining schedules with split dW/dI because the dW and dI are called kind of independently and it may be hard to find a place to have a single context region.

File: torch/utils/checkpoint.py (hunk 6 above)
Comment (soulitzer) [reply torch/utils/checkpoint.py]: Getting rid of the counter entirely. Now we're keyed on the GraphExecGroup object itself. 
Also made the state thread local.

Comment (soulitzer) [reply torch/utils/checkpoint.py]: Tests are failing because we probably need to add it to the tls state to so that autograd can ferry it to backward.

Comment (soulitzer) [reply torch/utils/checkpoint.py]: Updated now.

File: torch/utils/checkpoint.py (hunk 4 above)
Comment (soulitzer) [reply torch/utils/checkpoint.py:1610]: (non-reentrant has been updated in the description)

File: c10/core/AutogradState.h [hunk 8]
```diff
@@ -61,13 +68,18 @@ struct C10_API AutogradState {
...
+  }
+
  private:
   bool grad_mode_ : 1;
   bool inference_mode_ : 1;
   bool fw_grad_mode_ : 1;
   bool multithreading_enabled_ : 1;
   // NOLINTNEXTLINE(cppcoreguidelines-use-default-member-init)
   bool view_replay_enabled_ : 1;
+  std::optional<SafePyObject> graph_exec_group_;
```
Comment (albanD) [inline c10/core/AutogradState.h]: Move above the boolean field to prevent extra padding

**Review by albanD — APPROVED** [review]
Small thin, but otherwise good to land!
